gcm/gcm.py

```diff
@@ -263,79 +266,49 @@ def extract_unsent_reg_ids(self, info):
 
         :param registration_id: string of the registration id
         :param data: dict mapping of key-value pairs of messages
         :return dict of response body from Google including multicast_id, success, failure, canonical_ids, etc
-        :raises GCMMissingRegistrationException: if registration_id is not provided
         """
 
-        if not registration_id:
-            raise GCMMissingRegistrationException("Missing registration_id")
-
-        payload = self.construct_payload(
-            registration_id,
-            data=data,
-            collapse_key=collapse_key,
-            delay_while_idle=delay_while_idle,
-            time_to_live=time_to_live,
-            is_json=False,
-            dry_run=dry_run,
-            restricted_package_name=restricted_package_name,
-            priority=priority,
-        )
+        kwargs['is_json'] = False
+        retries = kwargs.pop('retries', 5)
+        payload = self.construct_payload(**kwargs)
 
         attempt = 0
         backoff = self.BACKOFF_INITIAL_DELAY
-        for attempt in range(retries):
-            try:
-                response = self.make_request(payload, is_json=False)
-                return self.handle_plaintext_response(response)
-            except GCMUnavailableException:
-                sleep_time = backoff / 2 + random.randrange(backoff)
-                time.sleep(float(sleep_time) / 1000)
-                if 2 * backoff < self.MAX_BACKOFF_DELAY:
-                    backoff *= 2
+
+        if retries:
+            for attempt in range(retries):
+                try:
+                    response = self.make_request(payload, is_json=False)
+                    return self.handle_plaintext_response(response)
+                except GCMUnavailableException:
+                    sleep_time = backoff / 2 + random.randrange(backoff)
+                    time.sleep(float(sleep_time) / 1000)
+                    if 2 * backoff < self.MAX_BACKOFF_DELAY:
+                        backoff *= 2
 
         raise IOError("Could not make request after %d attempts" % attempt)
 
-    def json_request(self, registration_ids, data=None, collapse_key=None,
-                     delay_while_idle=False, time_to_live=None, retries=5,
-                     dry_run=False, restricted_package_name=None, priority=None):
+    def json_request(self, **kwargs):
         """
         Makes a JSON request to GCM servers
 
         :param registration_ids: list of the registration ids
         :param data: dict mapping of key-value pairs of messages
         :return dict of response body from Google including multicast_id, success, failure, canonical_ids, etc
-        :raises GCMMissingRegistrationException: if the list of registration_ids is empty
-        :raises GCMTooManyRegIdsException: if the list of registration_ids exceeds 1000 items
         """
-
-        if not registration_ids:
-            raise GCMMissingRegistrationException("Missing registration_ids")
```

Comment:
If we remove the checking on the client I think we should handle the check in raise_error method when the response comes back.